### PR TITLE
auto reporting

### PR DESCRIPTION
**Pull request by wabinyai**
**_WHAT DOES THIS PR DO?_**
This PR implements a new API endpoint for generating comprehensive air quality reports with spatial analysis.
The endpoint supports both LLaMA and Gemini models for report generation, providing detailed analysis of air quality data across specified grid locations and time periods.

Added environment variables:
ANALTICS_URL=https://platform.airqo.net/api/v2/analytics/grid/report
GOOGLE_API_KEY=
HUGGING_FACE_TOKEN=
OPENAI_API_KEY=
To test the code, please have GOOGLE_API_KEY https://ai.google.dev/gemini-api

[Provide a brief description of the changes made in this PR]
There are 3 API to test 

-  `http://127.0.0.1:5000/api/v2/spatial/air_quality_report`

```
{
    "grid_id": "654230870a765d0013ec2bbd",
    "start_time": "2024-01-01T00:00",
    "end_time": "2024-01-27T00:00",
    "audience":"general public"
}
```
This uses gemini LLM , there are three types of audience
general public, policymaker, and researcher
 
-  ` http://127.0.0.1:5000/api/v2/spatial/air_quality_report_without_llm`
This uses rule base
```
{
    "grid_id": "654229088f9ef500139a24d8",
    "start_time": "2024-01-01T00:00",
    "end_time": "2024-01-27T00:00" 
} 
```

-  `http://127.0.0.1:5000/api/v2/spatial/air_quality_report_with_customised_prompt`
```
{
    "grid_id": "654229088f9ef500139a24d8",
    "start_time": "2024-01-01T00:00",
    "end_time": "2024-01-27T00:00",
    "customise_prompt":"what are top 3 most poluted sites and the pm2.5 value"
}
```
This API uses gemini to act as a chat tool

**_WHAT ISSUES ARE RELATED TO THIS PR?_**

- Jira cards
    - []
- GitHub issues
    - Closes #<enter GitHub issue number here>

**_HOW DO I TEST OUT THIS PR?_**

- [ ] Analytics: [Link to Analytics README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/analytics#readme)
- [ ] Auth Service: [Link to Auth Service README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/auth-service#readme)
- [ ] Calibrate: [Link to Calibrate README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/calibrate#readme)
- [ ] Data Proxy: [Link to Data Proxy README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/data-proxy#readme)
- [ ] Device Monitoring: [Link to Device Monitoring README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/device-monitoring#readme)
- [ ] Device Registry: [Link to Device Registry README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/device-registry#readme)
- [ ] Device Status: [Link to Device Status README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/device-status#readme)
- [ ] Device Uptime: [Link to Device Uptime README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/device-uptime#readme)
- [ ] Exceedances: [Link to Exceedances README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/exceedances#readme)
- [ ] Firebase: [Link to Firebase README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/firebase)
- [ ] GP Model: [Link to GP Model README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/gp-model#readme)
- [ ] Incentives: [Link to Incentives README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/incentives#readme)
- [ ] Kafka Connectors: [Link to Kafka Connectors](https://github.com/airqo-platform/AirQo-api/tree/staging/src/kafka-connectors)
- [ ] Locate: [Link to Locate README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/locate#readme)
- [ ] Metadata: [Link to Metadata README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/meta-data#readme)
- [ ] Predict: [Link to Predict README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/predict#readme)
- [ ] Spatial: [Link to Spatial](https://github.com/airqo-platform/AirQo-api/tree/staging/src/spatial)
- [ ] View: [Link to View README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/view#readme)
- [ ] Workflows: [Link to Workflows README](https://github.com/airqo-platform/AirQo-api/tree/staging/src/workflows#readme)

**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:_** 

- [ ] Endpoint 1
- [ ] Endpoint 2

**_ARE THERE ANY RELATED PRs?_**

- [ ] Related PR 1
- [ ] Related PR 2

## Release Notes

- **New Features**
  - Introduced new API keys configuration for Analytics, Hugging Face, Google, and OpenAI.
  - Added three new endpoints for generating air quality reports, enhancing user access to report generation.
  - Implemented a data fetching class to retrieve and process air quality data from external APIs.
  - Created a report generation class that produces tailored reports for researchers, policymakers, and the general public.

- **Dependencies**
  - Updated project dependencies to include libraries for machine learning and natural language processing, including `torch`, `transformers`, and `openai`.

### CODE REVIEW

File: src/spatial/models/report_datafetcher.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+class DataFetcher:
+    @staticmethod
+    def fetch_air_quality_data_a(grid_id, start_time, end_time):
+        token = Config.AIRQO_API_TOKEN  
+        if token is None:
+            print("Error: AIRQO_API_TOKEN environment variable is not set.")
+            return None
+
+        url = f"https://platform.airqo.net/api/v2/analytics/grid/report?token={token}"
```
Comment (Baalmart) [inline src/spatial/models/report_datafetcher.py:27]: @wabinyai , is it possible to have this URL (or a section of it) as an environmental variable? Thanks for the great work.

**Review by Baalmart — COMMENTED** [review]
Thanks @wabinyai , I have just left one small (minor) comment. Also:

1. Please let me (us) know if you have any new ENVs that need to be added.
2. Also, please let me (us)  know when this PR is ready for merging. 

Thanks again.

File: src/spatial/models/report_datafetcher.py (hunk 1 above)
Comment (wabinyai) [reply src/spatial/models/report_datafetcher.py:27]: @Baalmart I am still doing some extra work on it. 
The environment needed are **HUGGING_FACE_TOKEN**  **GOOGLE_API_KEY**   **OPENAI_API_KEY**
We are still exploring  other free LLM that can do as well as gemini

**Review by Baalmart — APPROVED** [review]
Thank you @wabinyai . Thanks for cooking! :)

**Pull request merged**